File: examples/primitives/python/src/operators/counter_pygen.py
from __future__ import annotations
from typing import *
import examples.operators.counter as pybind_counter_pygen
from bind_utils import *
from metaclass import *


class Counter(metaclass=OriginalMethodsMetaclass):
    """
    An example for with overloaded methods.
    """
    @bind
    def __init__(self, count: int):
        pass


    @property
    @bind
    def count(self) -> int:
        """
        Getter for count."""
        pass



    @bind
    def __add__(self, counter: Counter) -> Counter:
        """
        Plus operator"""
        pass
    @bind
    def compare_to(self, counter: Counter) -> int:
        """
        Comparison operator for kotlin"""
        pass
    @bind
    def __gt__(self, counter: Counter) -> bool:
        """
        """
        pass
    @bind
    def __iadd__(self, counter: Counter) -> Counter:
        """
        """
        pass

File: examples/primitives/python/wrappers/operators/counter_pygen.cpp
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/functional.h>

#include "cxx/operators/counter.hpp"

namespace py = pybind11;

void bindCounter(py::module& m) {
    using namespace iegen::example;
    py::class_<iegen::example::Counter> counter(m, "Counter");

    counter.def(py::init<int>(), py::arg("count"));


    counter.def_property_readonly("count", &iegen::example::Counter::count);
    counter.def("__add__", &iegen::example::Counter::operator+, py::arg("counter"));
    counter.def("compare_to", &iegen::example::Counter::compareTo, py::arg("counter"));
    counter.def("__gt__", &iegen::example::Counter::operator>, py::arg("counter"));
    counter.def("__iadd__", &iegen::example::Counter::operator+=, py::arg("counter"));
}